File: test/sample.lua
-- single-line comment
--[[
  block comment
  @param name string
]]

local M = {}

local MAX_RETRIES = 3          -- constant
local greeting    = "hello"    -- string
local count       = 42         -- number
local ratio       = 3.14       -- float
local is_ready    = true       -- boolean
local nothing     = nil        -- nil builtin

--- Returns a greeter function for `name`.
---@param name string
---@return function
local function make_greeter(name)
  return function()
    return string.format("%s, %s!", greeting, name)
  end
end

M.greet = make_greeter("world")

-- Table / field / property
local config = {
  retries  = MAX_RETRIES,
  verbose  = false,
  callback = M.greet,
}

-- Control flow
for i = 1, config.retries do
  if i % 2 == 0 then
    io.write(tostring(i) .. "\n")
  elseif i == config.retries then
    error("max retries reached")
  else
    coroutine.yield()
  end
end

-- Class-like pattern with metatables
local Animal = {}
Animal.__index = Animal

---@param name string
---@param sound string
function Animal.new(name, sound)
  return setmetatable({ name = name, sound = sound }, Animal)
end

function Animal:speak()
  return string.format("%s says %s", self.name, self.sound)
end

local cat = Animal.new("Cat", "meow")
print(cat:speak())

-- Operator mix
local result = (count * 2 + 1) / ratio
local flag   = is_ready and not (result > 100)

-- String escape
local escaped = "line1\nline2\ttabbed"
local raw     = [[no \n escaping here]]

return M

File: test/sample.py
# single-line comment

"""
Module docstring.
:param value: the input value
"""

from __future__ import annotations
import os
import sys
from typing import Optional, Union
from dataclasses import dataclass, field

MAX_RETRIES: int = 3
GREETING: str    = "hello"


@dataclass
class Config:
    """Application configuration."""

    retries: int  = MAX_RETRIES
    verbose: bool = False
    tags: list[str] = field(default_factory=list)


class Animal:
    """Represents an animal."""

    species: str = "unknown"

    def __init__(self, name: str, sound: str) -> None:
        self.name  = name
        self.sound = sound

    def speak(self) -> str:
        return f"{self.name} says {self.sound!r}"

    @classmethod
    def from_dict(cls, data: dict) -> "Animal":
        return cls(data["name"], data["sound"])

    @staticmethod
    def is_valid(name: str) -> bool:
        return bool(name and name.strip())

    def __repr__(self) -> str:
        return f"Animal(name={self.name!r})"


def make_greeter(name: str):
    """Higher-order function returning a greeter."""
    greeting = f"Hello, {name}!"

    def greet() -> str:
        return greeting

    return greet


# Control flow
def process(value: Optional[int]) -> Union[int, None]:
    if value is None:
        return None
    elif value < 0:
        raise ValueError(f"negative value: {value}")

    for i in range(MAX_RETRIES):
        if i % 2 == 0:
            continue
        print(i)

    result = [x * 2 for x in range(value) if x % 2 == 0]

    match value:
        case 0:
            return 0
        case int(n) if n > 100:
            return n // 2
        case _:
            return value

    return sum(result)


# String escapes and raw strings
escaped = "line1\nline2\ttabbedA"
raw     = r"no \n escaping here"
multi   = """
    multi
    line
"""

# Builtins
items = list(range(10))
total = sum(items)
names = list(map(str, items))
evens = list(filter(lambda x: x % 2 == 0, items))

if __name__ == "__main__":
    cat = Animal("Cat", "meow")
    print(cat.speak())
    sys.exit(os.EX_OK)
